### ISSUE TITLE
Add support for consumer version selectors for deployed and released versions

### ISSUE THREAD

**Issue by bethesque** [pact-foundation/pact-net#311]
Please add support for the following keys to be used in the consumer version selectors when fetching pacts for verification:

```
{ "deployedOrReleased": true }

{ "deployed": true }

{ "released": true }

{ "environment": "<env name>" }
```

These keys can be used in various combinations ( eg. `{ "environment": "prod", released: true, consumer: "Foo" }` ). Please allow any combination of keys, and do not validate them locally, as the validation rules can change over time as support is added for new selectors - just ensure that any errors returned from the API are displayed clearly to the user.

See https://github.com/pact-foundation/pact_broker/blob/master/lib/pact_broker/doc/views/provider-pacts-for-verification.markdown for the usage docs of each selector.

As more selectors will be added over time, it's my preference that the consumer version selectors are not strongly typed, and are just passed through straight from the configuration object to the API, so that we don't have to make code changes to the Pact clients each time we add new selectors. (eg. there will be new selectors coming for branch support very soon). I understand that this is not compatible with the idioms of every Pact client language however, so do whatever you need to do to make the new selectors work elegantly within your language.

Please update https://docs.pact.io/pact_broker/advanced_topics/consumer_version_selectors/ for your language once the selectors have been released.

**Comment by adamrodger** [comment]
Is there a typo on the above example tags? Both released and deployed are on there twice, and the example in the paragraph below has `consumer` but that's not in the example block above...?

**Comment by adamrodger** [comment]
Also, can the same tag be specified multiple times, e.g. to get multiple environments?

**Comment by bethesque** [comment]
> Both released and deployed are on there twice.

My bad - I've updated the examples.  

> example in the paragraph below has consumer but that's not in the example block above

I didn't realise the consumer key was not already supported in pact net. Can you please add that too. You'll find all the possible keys here: https://github.com/pact-foundation/pact_broker/blob/master/lib/pact_broker/doc/views/provider-pacts-for-verification.markdown Please check if there are any missing.

> Also, can the same tag be specified multiple times, e.g. to get multiple environments?

You'd specify a separate consumer version selector for each environment.

```
{ deployed: true, environment: "production" }, { deployed: true, environment: "test" }
```

99% of the time, I recommend and expect people to use ` "deployedOrReleased": true }`, which does all the environments and all the deployments/releases.

**Comment by bethesque** [comment]
What is the status on this please? I'm updating the docs with a section on which languages/versions support the selector, so people know if they can use it or not.

**Comment by adamrodger** [comment]
I think it would have to go in v4 so it's probably a little way off.

At the moment I've not even looked into it because the requirement looks
complicated at first glance so I've not had the time to go through it yet.

**Comment by adamrodger** [comment]
@bethesque I've done a bit of checking up and it appears maybe we already support this? it's hard to tell.

At the moment in v4 we support everything that the `pact_verifier_cli` supports, and that has a `--consumer-version-selectors` argument which just takes a comma-separated string.

The problem is that there's no documentation on what options you can pass to that flag. Like is it a string of JSON matching the requirements in this ticket, or is just a list of tags to fetch? If it's the former then we already support this I think. If it's the latter then we'd need Rust FFI changes to support this feature.

**Comment by bethesque** [comment]
It shouldn't be a lot of work, because as far as I know, pact.net already supports the consumer version selectors. The only reason there would be a client change is that some of the clients use a class to represent the selector, rather than just generic hashes/json (I recommend the generic hash approach for this very reason - it means there's no client update required when a new selector comes out). I understand the latest rust FFI does already support the new selectors (cc: @mefellows). I don't know how to pass in the selectors, or why they'd be comma separated though.

**Comment by adamrodger** [comment]
This has been implemented in v4.0.0 and there are currently no plans to backport to v3.x or earlier
